enterprise_access/apps/content_assignments/migrations/0021_backfill_allocated_at.py

```diff
@@ -0,0 +1,85 @@
+"""
+Data migration to backfill the ``LearnerContentAssignment.allocated_at`` field.
+"""
+
+from django.db import migrations
+from django.utils import timezone
+
+from enterprise_access.apps.content_assignments.constants import AssignmentActions
+
+
+def forwards_func(apps, schema_editor):
+    """
+    Populates the ``allocated_at`` field for all assignments whose value of that field is null.
+    """
+    LearnerContentAssignment = apps.get_model('content_assignments', 'LearnerContentAssignment')
+    HistoricalLearnerContentAssignment = apps.get_model('content_assignments', 'HistoricalLearnerContentAssignment')
+    LearnerContentAssignmentAction = apps.get_model('content_assignments', 'LearnerContentAssignmentAction')
+
+    records_to_backfill = LearnerContentAssignment.objects.filter(
+        allocated_at=None,
+    )
+
+    records_to_save = []
+    historical_records_to_save = []
+
+    for assignment_record in records_to_backfill:
+        last_notify_action = LearnerContentAssignmentAction.objects.filter(
+            assignment=assignment_record,
+            action_type=AssignmentActions.NOTIFIED,
+            error_reason=None,
+        ).first()
+        if not last_notify_action:
+            allocation_time = assignment_record.created
+        else:
+            allocation_time = last_notify_action.completed_at
+
+        assignment_record.allocated_at = allocation_time
+        assignment_record.modified = timezone.now()
+        records_to_save.append(assignment_record)
+
+        # Note: The reason we need to manually create historical objects is that Django's bulk_create() built-in does not
```

Comment:
nit: should this comment reference `bulk_update` as opposed to `bulk_create` since `bulk_update` is used to update existing `LearnerContentAssignment` objects, not creating new records? Same principle applies, though.